src/identity_committer.rs

```diff
@@ -0,0 +1,204 @@
+                        .await?;
+                }
+
+                select! {
+                    _ = wake_up_receiver.recv() => {
+                        debug!("Woke up by a request.");
+                    }
+                    _ = shutdown_receiver.recv() => {
+                        info!("Woke up by shutdown signal, exiting.");
+                        return Ok(());
+                    }
+                }
+            }
+        });
+        *instance = Some(RunningInstance {
+            handle,
+            wake_up_sender,
+            shutdown_sender,
+        });
+    }
+
+    #[instrument(level = "info", skip_all)]
+    async fn commit_identity(
+        database: &Database,
+        contracts: &Contracts,
+        tree_state: &TreeState,
+        group_id: usize,
+        commitment: Hash,
+    ) -> eyre::Result<()> {
+        // Get a progress lock on the tree for the duration of this operation. Holding a
+        // progress lock ensures no other thread calls tries to insert an identity into
+        // the contract, as that is an order dependent operation.
+        let tree = tree_state.merkle_tree.progress().await.map_err(|e| {
+            error!(?e, "Failed to obtain tree lock in commit_identity.");
+            e
+        })?;
+
+        // Fetch next leaf index
+        let identity_index = tree_state.next_leaf.fetch_add(1, Ordering::AcqRel);
+
+        // Send Semaphore transaction
+        let receipt = contracts.insert_identity(commitment).await.map_err(|e| {
+            error!(?e, "Failed to insert identity to contract.");
+            e
+        })?;
+
+        let mut tree = tree.upgrade_to_write().await.map_err(|e| {
+            error!(?e, "Failed to obtain tree lock in insert_identity.");
+            e
+        })?;
+
+        // Update  merkle tree
+        tree.set(identity_index, commitment);
+
+        // Downgrade write lock to progress lock – tree is now up to date, but still
+        // need to update the database.
+        let tree = tree.downgrade_to_progress();
+
+        let block = receipt
+            .block_number
+            .expect("Transaction is mined, block number must be present.");
+        info!(
+            "Identity inserted in block {} at index {}.",
+            block, identity_index
+        );
+        database
+            .mark_identity_inserted(group_id, &commitment, block.as_usize(), identity_index)
+            .await?;
+
+        // Check tree root
+        contracts
+            .assert_valid_root(tree.root())
+            .await
+            .map_err(|error| {
+                error!(
+                    computed_root = ?tree.root(),
+                    ?error,
+                    "Root mismatch between tree and contract."
+                );
+                error
+            })?;
+
+        // Immediately write the tree to storage, before anyone else can
+        // TODO: Store tree in database
```

Comment:
I think this comment is copied over, but no longer relevant? At any rate the TODO is captured in one of the open issues, so we can remove it here.